src/java-interop/java-interop-jvm.c

```diff
@@ -0,0 +1,60 @@
+#include <stdlib.h>
+#include <dlfcn.h>
+
+#include "java-interop-jvm.h"
+#include "java-interop-logger.h"
+#include "java-interop-util.h"
+
+static struct DylibJVM *jvm = NULL;
+
+int java_interop_jvm_load (const char *path)
+{
+	jvm = calloc (1, sizeof (struct DylibJVM));
+
+	jvm->dl_handle = dlopen (path, RTLD_LAZY);
+
+	if (!jvm->dl_handle)
+		return 0;
+
+	int symbols_missing = 0;
+
+#define LOAD_SYMBOL(symbol) \
+	jvm->symbol = dlsym (jvm->dl_handle, #symbol); \
+	if (!jvm->symbol) { \
+		log_error (LOG_DEFAULT, "Failed to load JVM symbol: %s", #symbol); \
+		symbols_missing = 1; \
+	}
+
+	LOAD_SYMBOL(JNI_CreateJavaVM)
+	LOAD_SYMBOL(JNI_GetCreatedJavaVMs)
+
+	if (symbols_missing) {
+		log_fatal (LOG_DEFAULT, "Failed to load some Mono symbols, aborting...");
+		exit (FATAL_EXIT_JVM_MISSING_SYMBOLS);
+	}
+
+	return 1;
+}
+
+static inline void
+_assert_dl_handle ()
+{
+	if (!jvm || !jvm->dl_handle) {
+		log_fatal (LOG_DEFAULT, "Missing JVM symbols!");
+		exit (FATAL_EXIT_JVM_MISSING_SYMBOLS);
+	}
+}
+
+int java_interop_jvm_create (JavaVM **p_vm, void **p_env, void *vm_args)
+{
+	_assert_dl_handle ();
+
+	return (*jvm->JNI_CreateJavaVM) (p_vm, p_env, vm_args);
+}
+
+int java_interop_jvm_list (JavaVM **vmBuf, int bufLen, int *nVMs)
+{
+	_assert_dl_handle ();
```

Comment:
...though my newly stated "we should stop using `jvm!JVM_GetCreatedJavaVMs()` entirely!" logic likewise means that we shouldn't call `_assert_dl_handle()` here and elsewhere.

Instead, we should return an error code:

```c
#define return_unless_dl_handle()  do { if (jvm == NULL) return -2; } while (0)

/* ... */
int
java_interop_jvm_list (JavaVM **vmBuf, int bufLen, int *nVMs)
{
    return_unless_dl_handle ();
    return (*jvm->JNI_GetCreatedJavaVMs) (vmBuf, bufLen, nVMs);
}
```

Reply:
That seems weird. It will behave differently than mono's Dylib. In the beginning I wanted to have similar layer to mono's dylib mechanism. Now it is going in different direction, where instead of having simple layer to access native methods from jvm, we are adding a logic here and will have to handle the error on higher levels.

Reply:
What if we need to add another function with different return type?

Reply:
> What if we need to add another function with different return type?

That's why mono/eglib uses a bunch of `g_return_val_if_fail()` macros...  Same basic idea.

So we could instead do:

```c
#define ji_return_val_if_fail(expr, val)  do { if (!(expr)) return (val); } while (0)
#define ji_return_val_unless_dl_handle(val) ji_return_val_if_fail(jvm == NULL, val)

/* ... */
int
java_interop_jvm_list (JavaVM **vmBuf, int bufLen, int *nVMs)
{
    ji_return_val_unless_dl_handle (-2);
    return (*jvm->JNI_GetCreatedJavaVMs) (vmBuf, bufLen, nVMs);
}
```

Reply:
I still don't like it. The intention of the PR was to make `jvm` functions available in the desktop environment and not change the behavior of them. I like to have the changes small and simple. If we need to change the logic (do we?), then we might do it in another PR?

The assert here would make it fail in the same place as before, just with better information why. If we add and return some error value, we have to handle it somewhere else. Note that the assert is normally not triggered as we check whether the library (and its symbols) were loaded and throw exception when creating the runtime.